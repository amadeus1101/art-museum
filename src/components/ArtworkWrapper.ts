import { colors } from '../constants/colors';
import { devices } from '../constants/devices';
import styled from 'styled-components';

export const ArtworkWrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  .illustration {
    width: 497px;
    aspect-ratio: 497/570;
    margin-right: 32px;
    position: relative;
    overflow: hidden;
    img {
      width: 100%;
      position: absolute;
      left: 0;
      top: 0;
      object-fit: cover;
    }
    .bookmark__wrapper {
      z-index: 90;
      position: absolute;
      right: 16px;
      top: 16px;
      img {
        width: auto;
        height: auto;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
      }
    }
  }
  .content {
    max-width: 640px;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    h4 {
      font-family: 'LexendRegular', sans-serif;
      font-size: 32px;
      color: ${colors.black};
    }
    .legend {
      p {
        margin-top: 32px;
        font-family: 'LexendRegular', sans-serif;
        font-size: 24px;
        color: ${colors.orange_light};
      }
      b {
        margin-top: 16px;
        display: block;
        font-family: 'LexendBold', sans-serif;
        font-size: 16px;
        color: ${colors.black};
      }
    }
    .overview {
      margin-top: 32px;
      ul {
        margin-top: 32px;
        li {
          margin-top: 16px;
          p {
            margin-right: 10px;
            font-family: 'LexendRegular', sans-serif;
            font-size: 16px;
            color: ${colors.orange_light};
          }
          span {
            font-family: 'LexendRegular', sans-serif;
            font-size: 16px;
            color: ${colors.black};
          }
        }
      }
    }
  }
  @media ${devices.laptop} {
    .content {
      h4 {
        font-size: 24px;
      }
      .legend {
        p {
          margin-top: 24px;
          font-size: 18px;
        }
        b {
          margin-top: 12px;
          font-size: 14px;
        }
      }
      .overview {
        ul {
          margin-top: 24px;
          li {
            margin-top: 12px;
            p,
            span {
              font-size: 14px;
            }
          }
        }
      }
    }
  }
  @media ${devices.tablet} {
    flex-direction: column;
    align-items: center;
    .illustration {
      max-width: 250px;
      margin: 0;
    }
    .content {
      .legend {
        margin-top: 32px;
        p {
          margin-top: 12px;
        }
        b {
          margin-top: 8px;
        }
      }
    }
  }
`;
