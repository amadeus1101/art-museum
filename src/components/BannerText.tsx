import { theme } from '../constants/theme';
import { devices } from '../constants/devices';

import styled from 'styled-components';

const BannerText = styled.h2`
  max-width: 684px;
  margin: 120px auto 72px;
  text-align: center;
  font-family: 'LexendBold', sans-serif;
  font-size: 64px;
  color: ${theme.colors.black};
  span {
    color: #f17900;
    white-space: nowrap;
    img {
      width: 74px;
      height: 74px;
    }
  }
  @media ${devices.tablet} {
    max-width: 513px;
    font-size: 48px;
    span img {
      width: 48px;
      height: 48px;
    }
  }
`;

export default BannerText;
