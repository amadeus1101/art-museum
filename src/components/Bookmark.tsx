import styled from 'styled-components';
import { theme } from '../constants/theme';
import bookmark_icon from '../assets/img/icon-bookmark-2.png';

const BookmarkStyles = styled.div`
  width: 59px;
  height: 59px;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #f9f9f9;
  border-radius: 50%;
  transition: 0.3s;
  &:hover {
    cursor: pointer;
    background: rgba(251, 215, 178, 30%);
  }
  img {
    width: 24px;
    height: 24px;
  }
`;

const Bookmark = () => {
  return (
    <BookmarkStyles>
      <img src={bookmark_icon} alt="bookmark-icon" />
    </BookmarkStyles>
  );
};

export default Bookmark;
