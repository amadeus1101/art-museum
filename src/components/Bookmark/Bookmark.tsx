import { FC } from 'react';

import { BookmarkStyles } from './styled';
import bookmark_icon from '../../assets/img/icon-bookmark-2.png';
import bookmark_active from '../../assets/img/bookmark-checked.png';

const Bookmark: FC<{ state: boolean; onClickBookmark: () => void }> = ({
  state,
  onClickBookmark
}) => {
  return (
    <BookmarkStyles $active={state} onClick={() => onClickBookmark()}>
      <img src={!state ? bookmark_icon : bookmark_active} alt="bookmark-icon" />
    </BookmarkStyles>
  );
};

export default Bookmark;
