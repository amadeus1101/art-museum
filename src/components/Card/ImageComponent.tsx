import { useState, useEffect, FC } from 'react';

type TImageProps = {
  src: string;
  placeholder: string;
  alt: string;
};

export const ImageComponent: FC<TImageProps> = ({ src, placeholder, alt }) => {
  const [imageSrc, setImageSrc] = useState(placeholder);
  useEffect(() => {
    const img = new Image();
    img.src = src;
    img.onload = () => {
      setImageSrc(src);
    };
    img.onerror = () => {
      setImageSrc(placeholder);
    };
  }, [src]);

  return <img src={imageSrc} alt={alt} loading="lazy" />;
};
