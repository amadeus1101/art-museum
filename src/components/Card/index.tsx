import { FC } from 'react';
import CardStyles from './styled';
import default_icon from '../../assets/img/image-default.png';
import Bookmark from '../Bookmark';

interface ICard {
  icon: string;
  title: string;
  author: string;
  status: string;
}

const Card: FC<ICard> = ({ icon, title, author, status }) => {
  return (
    <CardStyles>
      <div className="illustration">
        <img src={default_icon} alt="artwork-illustration" />
      </div>
      <div className="description">
        <div className="content">
          <p>{title}</p>
          <span>{author}</span>
          <b>{status}</b>
        </div>
        <Bookmark />
      </div>
    </CardStyles>
  );
};

export default Card;
