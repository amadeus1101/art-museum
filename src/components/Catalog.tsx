import axios from 'axios';
import { FC, useState, useEffect } from 'react';
import { CardType } from '@constants/CardType';
import { IFavourites } from '@constants/IFavourites';

import { CardsWrapper, Flex } from './CatalogStyles';
import { CardItemWrapper } from './Card/styled';
import Headline from './Headline';
import Card from './Card';

const Catalog: FC<IFavourites> = ({ favourites, callback }) => {
  const [cards, setCards] = useState<CardType[]>([]);

  useEffect(() => {
    async function getCards() {
      try {
        const cardsResp = await axios.get(
          `https://api.artic.edu/api/v1/artworks?fields=id,title,artist_title,is_public_domain,image_id&page=2&limit=18`
        );
        setCards(cardsResp.data.data);
      } catch (err) {
        console.log('Cards error: ', err);
      }
    }
    getCards();
  }, []);
  return (
    <CardsWrapper>
      <Headline title="Other works for you" subtitle="Here some more" />
      <Flex>
        {cards.map((card) => (
          <CardItemWrapper key={card.id}>
            <Card
              {...card}
              state={favourites?.find((elem) => elem.id === card.id) ? true : false}
              callback={callback}
            />
          </CardItemWrapper>
        ))}
      </Flex>
    </CardsWrapper>
  );
};

export default Catalog;
