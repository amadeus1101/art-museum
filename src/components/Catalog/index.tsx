import { FC, useState } from 'react';
import { FavouritesType } from '@constants/FavouritesType';
import { CardType } from '@constants/CardType';
import { useFetch } from '../../hooks/useFetch';
import { sortCards } from '../../utils/sortCards';

import CatalogPlaceholder from './placeholder';
import Headline from '../Headline';
import Card from '../Card';
import Sort from '../Sort';
import { Flex } from './styled';
import { CardItemWrapper } from '../Card/styled';

const Catalog: FC<FavouritesType> = ({ favourites, callback }) => {
	const [sortParam, setSortParam] = useState(0);
	const { data, loading, error } = useFetch<CardType[]>(
		'?fields=id,title,artist_title,is_public_domain,image_id&page=2&limit=18',
		'const'
	);
	if (loading) return <CatalogPlaceholder />;
	if (error)
		return (
			<Headline
				title="Cannot load catalog, try again"
				subtitle="Something gone wrong, we are really sorry;("
			/>
		);
	return (
		<>
			<Headline title="Other works for you" subtitle="Here some more" />
			<Sort param={sortParam} changeParam={setSortParam} />
			<Flex>
				{data &&
					sortCards(data, sortParam + 1).map((card) => (
						<CardItemWrapper key={card.id}>
							<Card
								{...card}
								state={
									favourites?.find((elem) => elem.id === card.id) ? true : false
								}
								callback={callback}
							/>
						</CardItemWrapper>
					))}
			</Flex>
		</>
	);
};

export default Catalog;
