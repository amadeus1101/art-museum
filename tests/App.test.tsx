import * as React from 'react';
import renderer from 'react-test-renderer';
import Headline from '../src/components/Headline';
import Logo from '../src/components/Logo';
import BannerText from '../src/components/BannerText';
import { ArtworkWrapper } from '../src/components/ArtworkWrapper';
import Bookmark from '../src/components/Bookmark';
import Footer from '../src/components/Footer';
import { Wrapper } from '../src/components/WrapperStyles';
import Card from '../src/components/Card';
import Header from '../src/components/Header';
import Pagination from '../src/components/Pagination';
import Sort from '../src/components/Sort';
import ErrorBoundary from '../src/components/ErrorBoundary';
import { ImageComponent } from '../src/components/Card/ImageComponent';
import Catalog from '../src/components/Catalog';
import Gallery from '../src/components/Gallery';
import Search from '../src/components/Search';
import Favourites from '../src/pages/Favourites';
import Artwork from '../src/pages/Artwork';
import Home from '../src/pages/Home';

import { preloadCards } from '../src/utils/preloadCards';
import { sortCards } from '../src/utils/sortCards';
import { fetchData } from '../src/utils/fetchData';

describe('Headline', function () {
	it('should display headline with data', function () {
		const component = renderer.create(
			<Headline title="Title" subtitle="Subtitle" />
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should display headline without data', function () {
		const component = renderer.create(<Headline title="" subtitle="" />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Logo', function () {
	it('should display logo1', function () {
		const component = renderer.create(<Logo icon={1} />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should display logo2', function () {
		const component = renderer.create(<Logo icon={0} />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('BannerText', function () {
	it('Should display h1', function () {
		const component = renderer.create(<BannerText>hi</BannerText>);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('ArtworkWrapper', function () {
	it('Should display css-wrapper for artwork', function () {
		const component = renderer.create(<ArtworkWrapper>styles</ArtworkWrapper>);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Bookmark', function () {
	it('Should display bookmark icon inactive', function () {
		const component = renderer.create(
			<Bookmark state={false} onClickBookmark={() => {}} />
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Should display bookmark icon active', function () {
		const component = renderer.create(
			<Bookmark state={true} onClickBookmark={() => {}} />
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Footer', function () {
	it('Should display footer', function () {
		const component = renderer.create(<Footer />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Pagination', function () {
	it('Should display pagination layout', function () {
		const component = renderer.create(
			<Pagination
				activePage="1"
				pages={['1', '2', '3', '4', '>']}
				onClickPage={() => {}}
			/>
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Sort', function () {
	it('Should display sort button', function () {
		const component = renderer.create(
			<Sort param={1} changeParam={() => {}} />
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Header', function () {
	it('Should display header', function () {
		const component = renderer.create(<Header />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Card', function () {
	it('Should display common card', function () {
		const component = renderer.create(<Card />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Wrapper', function () {
	it('Should display app style wrapper', function () {
		const component = renderer.create(<Wrapper>styles</Wrapper>);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('ErrorBoundary', function () {
	it('Should display error handler', function () {
		const component = renderer.create(<ErrorBoundary>errors</ErrorBoundary>);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('ImageComponent', function () {
	it('Should display image', function () {
		const component = renderer.create(
			<ImageComponent src="" placeholder="" alt="" />
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Search', function () {
	it('Should display search component', function () {
		const component = renderer.create(<Search />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Catalog', function () {
	it('Should display catalog widget', function () {
		const component = renderer.create(<Catalog />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('Gallery', function () {
	it('Should display gallery widget', function () {
		const component = renderer.create(<Gallery />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('HomePage', function () {
	it('Should display whole homepage', function () {
		const component = renderer.create(<Home />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('FavouritesPage', function () {
	it('Should display whole favouritespage', function () {
		const component = renderer.create(<Favourites />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('ArtworkPage', function () {
	it('Should display whole artworkpage', function () {
		const component = renderer.create(<Artwork />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('sortCards', function () {
	const tmp = [
		{
			id: 0,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},

		{
			id: 1,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 2,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 3,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 4,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
	];
	const result = [
		{
			id: 0,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 1,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 2,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 3,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 4,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
	];
	it('Should no sort cards', function () {
		expect(sortCards(tmp, 0)).toEqual(result);
	});
	it('Should sort cards array by a default order', function () {
		expect(sortCards(tmp, 1)).toEqual(result);
	});
	it('Should sort cards array by a title', function () {
		expect(sortCards(tmp, 2)).toEqual(result);
	});
	it('Should sort cards array by a author', function () {
		expect(sortCards(tmp, 3)).toEqual(result);
	});
});

describe('sortCards', function () {
	const tmp = [
		{
			id: 0,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},

		{
			id: 1,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 2,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 3,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
		{
			id: 4,
			title: 'Loading...',
			artist_title: 'Please wait',
			is_public_domain: false,
			image_id: '',
		},
	];
	it('Should sort cards array in order to condition', function () {
		expect(preloadCards(5)).toEqual(tmp);
	});
});
