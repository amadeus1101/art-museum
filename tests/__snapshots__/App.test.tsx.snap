// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtworkWrapper Should display css-wrapper for artwork 1`] = `
<div
  className="sc-hLQSwg hKHQnX"
>
  styles
</div>
`;

exports[`BannerText Should display h1 1`] = `
<h2
  className="sc-dmyCSP eHVAjv"
>
  hi
</h2>
`;

exports[`ErrorBoundary Should display error handler 1`] = `"errors"`;

exports[`Headline should display headline with data 1`] = `
<div
  className="sc-beySPh dkAqYt"
>
  <p>
    Subtitle
  </p>
  <h4>
    Title
  </h4>
</div>
`;

exports[`Headline should display headline without data 1`] = `
<div
  className="sc-beySPh dkAqYt"
>
  <p />
  <h4 />
</div>
`;

exports[`ImageComponent Should display image 1`] = `
<img
  alt=""
  loading="lazy"
  src=""
/>
`;

exports[`Pagination Should display pagination layout 1`] = `
<div
  className="sc-fsYfdN kqTWmG"
>
  <ul>
    <li
      className="active"
      onClick={[Function]}
    >
      1
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      2
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      3
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      4
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      &gt;
    </li>
  </ul>
</div>
`;

exports[`Sort Should display sort button 1`] = `
<div
  className="sc-qZrbh iKzYYv"
>
  <p>
    Sort by:
  </p>
  <span
    onClick={[Function]}
  >
    Title
  </span>
</div>
`;

exports[`Wrapper Should display app style wrapper 1`] = `
<div
  className="sc-gLLuof kUScnJ"
>
  styles
</div>
`;
